### ISSUE TITLE
Test parameter containing a semicolon

### ISSUE THREAD

**Issue by oznetmaster** [nunit/nunit#1885]
Currently, the --p (--params) command line option uses the semi-colon as a separator.  This makes it difficult to have a parameter containing a semi-colon,

I thought that using multiple --p option would work, but internally, all of the parameters are still  concatenated together using a semi-colon as a separator.

The external interface should not be changed.  However, the internal format (used to pass the option from the command line to the builder needs to be changed to use non-inputtable character as the separator.  Perhaps a null should be used?  Alternatively, since the options value is an object, the parameters could be passed as an array of strings?

Not sure why the decision was made to simple concatenate all the parameters separated by semicolons was made in the first place?

**Comment by CharliePoole** [comment]
This is basically the same issue as nunit/nunit-console#23, applied in this case to nunitlite. Since a lot of discussion has already taken place there, let's keep that as the primary issue and port the fix here when it's done.

**Comment by jnm2** [comment]
@CharliePoole I'm not sure that this issue should be specific to the console runner at all. I'm having this issue via the VSTest adapter. I send parameters to VSTest using a .runsettings file and anything after a semicolon is removed.

At least part of the issue is in this repo: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Api/DefaultTestAssemblyBuilder.cs#L138

This is an incredible pain because the only use case I have for parameters is connection strings... 😆

**Comment by CharliePoole** [comment]
@jnm2 I was aware of that and probably should have recorded it as such in the first place. I think any general solution will involve some sort of escaping by each individual runner and then interpretying that escaping in the DefaultTestAssemblyBuilder.

As is often the case, the NUnitLite solution is the easiest to work on first because the client is in the same repository as the framework.

BTW, any general solution will probably have to deal with escaping `=` as well as `;`.

**Comment by oznetmaster** [comment]
I am pretty sure this was fixed in nunitlite already.  Unless it was part of my PR that was not accepted, but I think not.

It should be checked.

**Comment by jnm2** [comment]
What work is left to do?

**Comment by CharliePoole** [comment]
Nothing has been done on this in nunitlite. Also, I don't it in PR #1863.

It's probably not quite the "easyfix" for a newbie, but should be easy enough for @jnm2 if you want to implement it. :smile:

**Comment by jnm2** [comment]
I'll get it done, it's the last thing blocking me right now.

**Comment by jnm2** [comment]
s/blocking/bugging

**Comment by jnm2** [comment]
I'd much rather leave up to the individual command line or adapter how to parse the parameters and whether to escape, and change `options[FrameworkPackageSettings.TestParameters]` to an `IDictionary<string, string>`.

**Comment by jnm2** [comment]
I also see tremendous value in a custom NUnitOptions type which exposes strongly typed properties like `TestParameters`. If necessary it could also implement `IDictionary<string, string>`.

**Comment by jnm2** [comment]
And here's the other loose end: https://github.com/nunit/nunit3-vs-adapter/blob/c8e99e6aefb64a0fb71d2ee2fb3405978aa4c766/src/NUnitTestAdapter/NUnitTestAdapter.cs#L193

**Comment by oznetmaster** [comment]
I proposed a fix for this quite a while ago, but it changes TestParameters to an IList, and it was considered a breaking change.

I changed it that way in my CF builds, and it work fine.  However, I am not constrained by the issue of breaking changes :(

**Comment by oznetmaster** [comment]
Also, there is no need to escape = because it can only occur on the RHS of the required =, and therefore is not ambiguous.

**Comment by jnm2** [comment]
@oznetmaster Yes, I read that in nunit/nunit-console#23.

What are all the things that would break if the type was changed to `IDictionary<string, string>`?

**Comment by oznetmaster** [comment]
Ask @CharliePoole.  I believe he was the one who indicated it would be an issue.

Also, in CommandLineOptions, there is no need for it to be an IDictionary, just an IList.

In TestContext, the TestParameters property is effectively an IDictionary already.

**Comment by oznetmaster** [comment]
I should note that I am only dealing with nunitlite.  I have never looked at the nunit.console code.

**Comment by jnm2** [comment]
`IDictionary` is superior in my mind because it enforces the invariant that no two keys are equal. Plus it's a fitting idiom.

**Comment by oznetmaster** [comment]
That would require that CommandLineOptions would now be imputing semantics to a command line option..

That has been the responsibility of the runner, and probably should stay there.

**Comment by CharliePoole** [comment]
@jnm2 I'd like to see the bug fixed. If you conflate fixing the bug with adding a new feature, that delays fixing the bug.

Delays are always bad, but 10 days from release they are especially bad.

Just sayin'

**Comment by jnm2** [comment]
@oznetmaster I think it's okay to impute dictionary semantics to `TestPackage.Settings["TestParameters"]`. That's exactly what I was hoping for because it would mean various CLIs are not coupled to a particular escaping scheme. Eventually we'd see things being unescaped one way and escaped again just for the runner.

@CharliePoole Me too. Is the break caused by making `TestPackage.Settings["TestParameters"]` an `IDictionary<string, string>` more than a few minutes' work?

**Comment by oznetmaster** [comment]
In fact, that is not really possible because of the fact that a semicolon is currently used to delimit multiple test parameters on the command line.

Therefore, a specific escaping scheme has to be specified in order to simply break the parameters apart.

I see no way to make this escaping scheme independent.

For instance, my CommandLineOptions uses the standard \ to escape semicolons that are not delimiters, and \\ to escape the backslash.

**Comment by oznetmaster** [comment]
Once the parameters are split correctly, any other interpretation of the parameter is really the responsibility of the runner.  The does, in fact, make it CLI independent.

The best scenario would be to not allow multiple test parameters per --p option, and then escaping would be only the responsibility of the runner.

**Comment by oznetmaster** [comment]
If an IDictionary is used, then is it case sensitive?  That should be the runners decisions.

What if the runner wants to actually allow multiple test parameters with the same name?  I ask, because my implementation actually does, and creates an array of values for the parameter.

**Comment by jnm2** [comment]
Okay, so clearly too much design would be involved before being able to work on an `IDictionary` solution. 
Just keep in mind for the future that the ideal scenario is where the VS test adapter hands an `IEnumerable<KeyValuePair<string, string>>` which it gets from VSTest and the runner uses that collection without any escaping. This is totally possible.

**Comment by oznetmaster** [comment]
That is very similar to using an IList.  But it does still require an escaping scheme for the CLI.

Another reason for not using an IDictionary is if the runner needs to know the order in which the parameters were specified.  Again, semantics vs syntax.

**Comment by jnm2** [comment]
`IEnumerable<KeyValuePair<string, string>>` answers all those needs.

**Comment by oznetmaster** [comment]
Yes, but why is it superior to simply passing an IList.  It means that the CLI, or test adapter, or GUI, still has to parse the parameter from the name=value into <name, value>if it does not have them in that form already.  A CLI will never have them in that form. The current GUI will probably not either, since the decision was made not to support parameter discovery.  Not sure how the VS test adapter would either, but it could.

My GUI, in fact, does optinally have them in that form, since I support parameter discovery.  I also use a table for the parameters, so even if they are entered manually, I have name and value already separated, and currently have to join them with an = to pass to my IList.

I think it is a case of six of one and a half-dozen of the other.  No matter which format is used, some front ends are going to have to do some parsing or conversion.

**Comment by oznetmaster** [comment]
As an extreme case, what if a runner whats to support unnamed parameters, that ares access as an array of values only?

IList supports that, IEnumerable<KeyValuePair<string, string>

**Comment by CharliePoole** [comment]
@jnm2 @oznetmaster I'm on my phone so a full response to all this will have to wait till I get home. One small thing: you've each talked about the "runner" in contexts where it's not clear what you mean. Remember that each client is called a runner, the engine contains a bunch of runners and the framework also has a runner. It's unfortunate that the term is so overloaded.

**Comment by jnm2** [comment]
@oznetmaster A CLI _should_ have them in separated key-value form. A CLI is responsible to parse `-p name=value` or `-p name value` as it sees fit. And the VSTest adapter certainly has them in that form handed in by VSTest.

**Comment by jnm2** [comment]
My point is, all front ends should be responsible for parsing. The NUnit framework DefaultTestAssemblyBuilder should not do any parsing. The VSTest adapter doesn't need parsing ebcause VSTest itself does the parsing. CLIs should do the parsing along with all their other arguments.

**Comment by jnm2** [comment]
@oznetmaster 

> As an extreme case, what if a runner whats to support unnamed parameters, that ares access as an array of values only?

On second thought, I stand by my `IDictionary<string, string>` idea. These parameters are always translated to a `NUnit.Framework.TestParameters` instance which *is* a `Dictionary<string, string>`, aren't they? They are meaningless otherwise, right?

Therefore, since there is no way to use parameters without going though the `NUnit.Framework.TestParameters` dictionary, that rules out case-insensitivity, it rules out multiple identical keys, and it rules out null keys.

**Comment by oznetmaster** [comment]
The fact that they are translated to a dictionary does not meant that there is a one to one correspondence between them!

In my CF build, there is not.




-

**Comment by oznetmaster** [comment]
I allow optional insensitivity, multiple to one (the array valued parameters), and a unnamed parameters that  map into a single unnamed array parameter.

In the latter two cases, order is very important, and the use of an IDictionary would loose the order.

**Comment by jnm2** [comment]
@oznetmaster, okay, but in your CF build how do you access the parameters? If you access them through `TestContext.Parameters` which is an `NUnit.Framework.TestParameters` instance, you can't have any of that.

**Comment by jnm2** [comment]
I propose a non-breaking change. If `options[FrameworkPackageSettings.TestParameters]` is a string, then business as usual and the string is parsed into the `TestContext.Parameters` dictionary. If it is `IDictionary<string, string>`, it feeds that straight into the `TestContext.Parameters` dictionary.

**Comment by oznetmaster** [comment]
Why do you think that is the case?

In fact, I support fully typed access to parameters, array access to parameters, and so on.

If you access Parameters property indexer (`Parameters[names]`) you get the string value.

But all uses I have seen reference the Get<T>(name, defaultValue) method, which allows for typed access to the parameters and for the use of default values.

T can be an array, and name can be null.

**Comment by oznetmaster** [comment]
Also, since the Parameters property uses an indexer, Parameters[null] or Parameters[String.Empty] is totally valid.

**Comment by oznetmaster** [comment]
I still disagree.

I like the non-breaking option of passing a string, but I think that the breaking option should be an IList, not an IDictionary, for all the reasons I have already listed.

**Comment by jnm2** [comment]
@oznetmaster I'm a bit worried that we're talking about two different things. NUnit's `DefaultTestAssemblyBuilder` takes the parameters and populates `TestContext.Parameters`. When you access `TestContext.Parameters[null]` you get an argument exception. Also, multiple values are simply overwritten with the last one.

**Comment by oznetmaster** [comment]
Not in my CF build.

I fixed the null "bug".

And since arrays valued parameters are supported, multiple values are appended to the end of the array.

Again, it is an issue of syntax vs symantics.

By forcing the parameters to be passed by an IDictionary, you are limiting how they can be used by a builder/runner.  If they are passed as a IList (as are many other command line options, like the list of tests), then there is no implicit limitation on how they are interpreted.

**Comment by oznetmaster** [comment]
BTW.. DefaultTestAssemblyBuilder appropriately shows the issue well.  It name starts with Default.  You want to define the semantics of ALL test assembly builders by the default one.

**Comment by jnm2** [comment]
@oznetmaster I really want to put eyes on this thing because I'm having trouble understanding what you're getting at. I want to see it in action. 

You're saying you are using array-valued parameters with and without names. Since I'm pretty sure NUnit doesn't support either of those things, I'm super curious why this is holding us up.

**Comment by jnm2** [comment]
So basically: you want your assembly builder to be able to *do* the parsing of the params command line from the CLI, just in case your assembly builder supports array-valued or unnamed parameters.

How does that even work if you have two `--params` switches specified at the command line? Already you're behavior-coupled to however the CLI front-end decides to join them.

**Comment by oznetmaster** [comment]
If my PR for that addition is accepted at some point in time, it will.

However, passing an IDictionary from any front end (GUI, CLI, VS adapter, ...) to a builder/runner is implicitly limiting in that it is not lossless.  It loses order and it can lose case insensitivity.

An IList loses nothing.

I am not sure why you are so hung up over the IDictionary?  It is one LINQ statement to convert the IList to and IDicitionary, if that is all you want to do.

**Comment by jnm2** [comment]
So basically: you want your assembly builder to be able to *do* the parsing of the params command line from the CLI, just in case your assembly builder supports array-valued or unnamed parameters. (Which still sounds far-fetched and I want to see it.)

How does that even work if you have two `--params` switches specified at the command line? Already you're behavior-coupled to however the CLI front-end decides to join them.
And how do you reconcile that with the fact that *there is no command line* with the VSTest adapter?

> I am not sure why you are so hung up over the IDictionary? It is one LINQ statement to convert the IList to and IDicitionary, if that is all you want to do.

Because with `IList<string>`, the assembly builder becomes responsible for parsing which part of the string is the key and which part is the value. There's no point in doing any of this if we're still going to leave the assembly builder responsible for any parsing.

**Comment by jnm2** [comment]
If multiple identical keys and null keys are a supported scenario by NUnit, `IEnumerable<KeyValuePair<string, string>>` will still resolve all of this. That is the minimum structure possible to not lose any information and to keep the assembly builder from having to do any parsing.

**Comment by oznetmaster** [comment]
No.  Two --params switches are not different from two --test switches.  They are passed as an IList.

As a matter of fact, there are many IList properties of CommandLineOptions : InputFiles, TestList, ResultOutputSpecifications, ExploreOutputSpecifications, ErrorMessages, 

Having TestParameters also be an IList is, in fact, totally consistent.  Having it be an IDictionary is an anomaly.

**Comment by oznetmaster** [comment]
I agreed with that a while ago,

However, it is still inconsistent with the exiting structure of CommandLineOptions, which has, to date, utilized IList to pass multiple valued options.

Let me give you one final scenario.  What if there was a later desire to add type information to the --params option.  e.g. `--params:(int)param1=123`.

IList can handle this.

**Comment by jnm2** [comment]
> No. Two --params switches are not different from two --test switches. They are passed as an IList.

Where is this happening? I don't see it. I see what seems to be the contrary: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/nunitlite/CommandLineOptions.cs#L181

**Comment by oznetmaster** [comment]
Yes, that is the bug that we are talking about fixing.

What I am saying is that if that fix is a collection (as it is in my CF build), it should be an IList to be consistent.

**Comment by oznetmaster** [comment]
Here is a radical proposal.  Allow all three possible values of options[FrameworkPackageSettings.TestParameters]

If it is a string (current) then it is parsed by the builder.

If it is an IList, then it is handled as desired by the builder.  The default builder can use the single LINQ statement to convert it to an IDictionary.

If it is an IDictionary, then it is used as is.

**Comment by jnm2** [comment]
Alright. Let's play this out. We have these parameter source formats:

Source Format 1: `IEnumerable<KeyValuePair<string, string>>` from VSTest .runsettings xml
Source Format 2: `IEnumerable<string>` holding the values of one or more CLI `--params` switches
Source Format 3: Maybe something unknown in the future, like typed objects from a property grid in a GUI.

**Scenario 1: For those using `DefaultTestAssemblyBuilder`:**

 * We all agree that Source Format 1 should go straight to `TestContext.Parameters`. This is the topic of this thread.
 * Source Format 2 or 3 will need to be parsed first into Source Format 1; otherwise it's useless to `DefaultTestAssemblyBuilder`. In my mind this parsing is the job of the front-end rather than the assembly builder, but that would prevent Scenario 2 from working which has been @oznetmaster's point.
 * Arguably the 99.99% scenario

**For those using a custom builder that uses something other than `TestContext.Parameters`:**

 * Source Format 1, 2 and 3 all need to be specially parsed for non-NUnit rules like multiple identical keys
 * Worrysome because I still thing that parsing Source Format 2 should be handled by the front-end itself so that it can have its own style `--params name=value` and `--params name value` rather than being coupled to the assembly builder's assumptions.
 * Arguably the 0.01% scenario. Perhaps NUnit should consider this an invalid scenario in the first place because NUnit only intends to support `TestContext.Parameters`. @CharliePoole?

----


In situations like this you want a common currency to restrict the cross product of combinations. What if `IDictionary<string, string>` were that currency? It fits the 99.99% case to a tee.

 * All front-ends would do their own parsing. `--params X 1` and `--params X=` would both return `new Dictionary { ["X"] = "1" }` at the CLI front-end's discretion.
 * Your custom assembly builder could still represent array-valued properties:
    - `--params X=1;X=2;X=3;Y=4` would not work, but `--params X=1|2|3;Y=4` would work. Your assembly builder would receive `new Dictionary { ["X"] = "1|2|3", ["Y"] = "4" }` and could easily do its own  custom splitting and unescaping.
    - `--params 1;2;3` would not work, but `--params =1|2|3` or `--params ""=1|2|3` would work. Your assembly builder would receive `new Dictionary { [""] = "1|2|3" }` and could proceed as it sees fit.
    - **Overall: no information lost**, you just have to work around the command line a little bit because the command line is now catering to the 99.99% case without blocking the 0.01% case
    - **Additional benefit** because the VSTest .runsettings would also magically get the ability to do `<property name="" value="1|2|3">` which the front-end would parse as `new Dictionary { [""] = "1|2|3" }` and your custom builder will automatically understand for free, with **no command line involved**!

**Comment by CharliePoole** [comment]
@jnm2 @oznetmaster I think you're taking a stroll down the garden path here. @oznetmaster has his own complete implementation of the "nunit" framework and "nunitlite." Since it only runs on a proprietary, modified version of the compact framework most of us can't see or exercise it. Where it's interesting is that we __could__ do something similar if we wanted to but only __if__ it didn't break anything else. The key difference is that we want to be able to work with many runners on many platforms and that requires an understanding of how the settings travel from layer to layer.

So... here's my attempt to explain what I think are the compatibility issues, in answer to one of @jnm2 's questions...

Runners (in the sense of client exes) get settings in various ways. If they are command-line programs, they get them from the command-line. Guis may get them from menus. Any may get them from a settings or config file. Additionally, an adapter like the NUnit 3 VS adapter may get settings from a controlling program (the IDE for example) that has pre-digested some of them. As @jnm2 points out, that's how we get parameters from Visual Studio.

CHANGES: It's trivial to change how the command-line works, provided nothing changes below it.

Runners pass settings to the engine using a TestPackage. This class is defined in the nunit.engine.api assembly and deliberately has a minimalist implementation. Since that assembly is unversioned (always 3.0.0.0 up to now) we try not to make changes to it. Within TestPackage, the Settings property is an `IDictionary<string, object>`. I considered changing it to `<string, string>` but instead chose to use a limited set of primitive types as objects. This could have been a design error, but it's not feasible to change it now. It's unfortunately necessary to know what types are accepted downstream as values in order to work with this. It's not as big a problem as it would be since all our options __except__ test parameters are predefined and since this is not directly accessible to users except for third-party developers.

CHANGES: We can't change the interfaces in `nunit.engine.api` except in an additive way. Even then, there has to be a way for clients to figure out what is available in the engine itself. That's handled by clients having the ability to demand a particular minumum engine version from the api.

If tests are being run in a separate process, the engine launches one of the agent exes. The agent contains a copy of the engine internally. TestPackages are passed to the subordinate engine from the primary engine using remoting. Similarly, when the are run in a particular domain, as they usually are, remoting is used to transfer the TestPackage.

CHANGES: All OK here, provided the settings values are remotable.

Eventually, all packages end up in the hands of a driver, whose purpose is to communicate with a particular framework. Currently, only `DirectTestRunner` communicates with a driver. It passes `TestPackage.Settings` directly to the driver's `Load` method. That implies that any drivers written, for any framework whatsoever, must accept this parameter. Again, nothing is stated about the values in the dictionary and each framework may have it's own restrictions on allowed types.

If we restrict ourselves to the NUnit 3 driver, which is what's relevant here, then we still have to bear in mind that the driver will call all versions of nunit.framework >= 3.0 in the same way. We could try to make the driver "smarter" and pass different things to different minor versions of the framework, but that's adding complexity we probably don't want in the driver. Ideally, we will keep using the same driver at least till there is an nunit 4.

CHANGES: Right now everything goes so long as the values are remotable. However, we have plans to provide alternate ways to with the framework where remoting will not serve. In fact @oznetmaster has already done that in his private build, I believe. The future-proofing limitation would be not use any Type we don't feel we could easily serialize and deserialize on our own - since we don't know what platform limitations we might face.

After all that, I have to say that I may have been wrong in suggesting nunitlite as a good place to start. It happens to be different from everything above since it doesn't use the engine and has a direct reference to the framework! It's probably the best place to start coding, but only after a more general discussion like this one.

**Comment by CharliePoole** [comment]
@jnm2 

I'm so far behind, I'd better quote you...

> Is the break caused by making TestPackage.Settings["TestParameters"] an IDictionary<string, string> more than a few minutes' work? The motivation for this is to avoid work building something at the wrong abstraction level (see the comment to oznetmaster). I don't think the VS adapter should be escaping.

Hard to say without trying it as a separate refactoring. My guess is that it's minimal __but__ that the corresponding change to the framework API might be the bigger risk.

**Comment by oznetmaster** [comment]
I did post most of the changes relating to TestParameters as a PR at one point in time, so they have all been/are fully visible.  This includes the ; fix, the IList, and so on, as well as many other things like discovery and validation.

**Comment by oznetmaster** [comment]
As a side note, my CF implementation is 99% source identical to the standard nunit framework and nunitlite.

As @CharliePoole  noted, this has been accomplished by providing the pieces missing from CF as part of the environment, rather than conditioning the source code to remove capabilities not supported by CF, as is done in the official CF builds.

This has allowed me to simply update my source to match the current builds/releases, with no additional conditioning required.  For the most part, this is an automated procedure.

**Comment by CharliePoole** [comment]
@jnm2 

> `IEnumerable<KeyValuePair<string, string>>` answers all those needs.

I agree, although I don't see why we wouldn't use an IDictionary. As @oznetmaster is pointing out, the use of `IEnumerable` doesn't guarantee order, which depends on the actual implementation class.

That said, I don't particularly want to guarantee order among the differently named parameters. It makes no sense to me. I do understand the desire for a particular parameter name to reference an array of values in a certain order, but that's another (future) thing.

Unless you have a good reason of your own to use `IEnumerable<KeyValuePair<string, string>>`, I'd go with <IDictionary<string, string>> for the test parameters. That's what they end up as in TestContext anyway.

> To be clear, CommandLineOptions.TestParameters stays a string.

Could you illuminate this a bit? Obviously, the user continues to type strings, but why wouldn't the property be a list of strings?

**Comment by jnm2** [comment]
@CharliePoole Good point about `IEnumerable` not guaranteeing order preservation nor do I care about order, but "1|2|3" preserves order when it is custom parsed by a custom assembly builder that cares about order. I prefer `IDictionary` as well.

> > To be clear, CommandLineOptions.TestParameters stays a string.

> Could you illuminate this a bit? Obviously, the user continues to type strings, but why wouldn't the property be a list of strings?

Ah yes. I meant, it stays as-is and does not become `IDictionary<string, string>`. If there are other reasons for wanting it to be `IList<string>`, they should be orthogonal to this discussion.

@CharliePoole what do you think of my [analysis comment](https://github.com/nunit/nunit/issues/1885#issuecomment-268680403)? Are you on board with `IDictionary`, seeing that custom extensions to NUnit have ways to continue doing what they need to around `IDictionary`?

What do you think about about my non-breaking change idea where DefaultTestAssemblyBuilder checks for `IDictionary<string, string>` before checking for `string`, so that front-ends can upgrade at their convenience?

**Comment by oznetmaster** [comment]
So make it an `IList<KeyValuePair<string, string>>`, which derives from IEnumerable and does guarantee order.  What way, you will have it if/when you want it.  And it can directly be converted to an IDictionary in the constructor if that is what you want for now.

**Comment by jnm2** [comment]
@oznetmaster You'll still have ordering for items in array-valued properties with IDictionary. Just not between properties with different names. Sounded like @CharliePoole doesn't think it's worth it.

**Comment by CharliePoole** [comment]
@jnm2 Hopefully, this note will cover all your pending questions. If not, just ask again...

You have me convinced that the engine should accept a dictionary value for TestParameters, except for compatibility issues, which I'll come back to. Of the current clients, the VS adapter has already pre-processed the arguments to the greatest extent and it doesn't seem logical that we should turn around and try to compress all that pre-processing into a simple string.

However, the engine has to deal with the 3.5 framework, which won't be able to handle such a value. Backward compatibility is real high on our scale of values, in particular as it relates to the engine dealing with older framework versions.

It seems to me that the simplest way to deal with this is to leverage the fact that consumers of the settings ignore any settings they don't know about. How about providing a package setting like "TestParameterDictionary" to hold this new value? The 3.6 framework could look for the new setting. If we want to provide __forward__ compatibility (to run the 3.6 framework with a 3.5 console runner, for example) we can do that by having the framework check for both values. I'm not sure what's best here. @rprouse what do you think?

If we go this route, we have to decide how to handle runners that use the old parameter setting. Should the engine create a dictionary for the runner? This relates somewhat to the answer we give to the earlier question about forward compatibility.

I'm a little lost in the part about "custom assembly builders." The framework was initially built with the capability of specifying custom assembly builders and runners as a way to easily move from one version of the builder to another. We haven't used it for a number of years and we never exposed it publicly (except there is an api that allows you to specify the Type name, provided such a Type is present in the framework). In fact we recently turned down a request to allow users to supply their own builders. Anyway, if that's what you are referring to, I don't think it's an issue.

-------

Looking back at the "forward compatibility" question 15 minutes later, I guess I can see some advantage in having the framework either use a passed in dictionary, or construct one. It wouldn't add a lot of code. I think I prefer having separate package settings rather than testing the Type of the object passed in.

**Comment by oznetmaster** [comment]
Just to be clear. My CF implementation already solves all these problems and more.

Your solution would be going backwards for me.

Having had all of the features I have mentioned in the field for over a year, I was just trying to impart the feedback I have gotten to your process.

I really do not care what you end up doing.  However, I always plan for the future.  And since I have implemented and deployed these "future" capabilities, I have an understanding of the underlying infrastructure required to support them.

I am convince if you use an IDictionary as the value passed in  TestPackage.Settings["TestParameters"] it will eventually become a limiting implementation unless the front end does all of the parsing and processing of the parameters.  Even then, expansion of the parameters to allow things like typing as per my example will not be possible.  These things may never happen.  However, why build in planned obsolescence for seeming no upside.

I am indifferent to what is decided in the end.

**Comment by CharliePoole** [comment]
@oznetmaster On the question of ordering... I understand why you might want a single parameter to have a value of an array or other ordered list. We can provide for that.

What I don't understand is why you want to maintain some sense of ordering across __different__ options. Let's say I invent a command-line where "--process:Multiple --domain:Multiple" has a different meaning from "--domain:Multiple --process:Multiple". I can't think of what that diffe.rence might be but in any case it seems like a very confusing user interface.

We have in the past discussed using positional options to apply to the assembly immediately preceding or following them. We haven't done it but it's at least conceivable. However, even if we did that, the options pertaining to a particular assembly would be in the subpackage for that assembly and order would not be relevant.

I think we should provide for future growth that we can reasonably expect. Can you give an example of something we might want to support?

**Comment by oznetmaster** [comment]
I am referring to ordering on the command line.

-p:param1:5 -p:param1:6 -p:param1:7

is passed (in my build) as an IList in TestPackage.Settings["TestParameters"], and then is processed into a single dictionary entry with a key of "param1" and a value of an array string "{5, 6, 7}" in my DefaultTestAssemblyBuilder.  It then can have its value obtained by any of the methods discussed previously.

This is really syntactic sugering masquerading as semantics, I guess :(

**Comment by oznetmaster** [comment]
Perhaps it should be noted that my GUI creates a command line and passes it to the remote nunit framework as a string.  So the command line is the only interface I currently use in the CF build.

**Comment by oznetmaster** [comment]
@CharliePoole  I never mentioned ordering across different options.  Just multiple occurrences of the same option.

**Comment by CharliePoole** [comment]
@oznetmaster Ordering among multiple params with the same name wasn't on my radar. AFAIK we don't support it right now, since it all gets converted to a dictionary. In fact, using that example, the order would be preserved but the insertion into the dictionary would fail.

However... let's say we decide to support multi-valued params as equivalent to an array. W e do that in test properties and could easily do it here. But even so, the single entry for param1={5,6,7} would not need to be in any particular order WRT the other entries.

> I never mentioned ordering across different options.  Just multiple occurrences of the same option.

Confused here! I originally thought you were talking about the use of a dictionary for the package Settings as a whole. Then I thought you meant for the value of the TestParameters setting, which applies to all --params options. But if you just mean all parameters of the same name, doesn't that imply processing by the runner... which I understand you don't think we should do?

@jnm2 has suggested there should be no processing in the adapter, because VS has already done the processing for us. I agree with that. I thought we all agreed that different runners with different user interfaces would have to create the dictionary (or other structure) in their own way.

**Comment by oznetmaster** [comment]
But doesn't making it an `IDictionary<string, string>` force a particular structure?

**Comment by oznetmaster** [comment]
I was not referring to order in regards to other parameters.  I was referring to maintain the order within itself.  The only way it could be passed as param1={5,6,7} would be if the aggregation were done in the front end (CLI in my case).  So now, the CLI would have to know how to convert multiple parameters with the same name into a single aggregate array parameter.  I wanted to put that functionality in the TestBuilder, which is where it is in my CF build.

**Comment by oznetmaster** [comment]
BTW, I do also support --param:param1={1,2,3} by default.  So does the current implementation of nunit.

**Comment by jnm2** [comment]
@oznetmaster

> But doesn't making it an `IDictionary<string, string>` force a particular structure?

Yes, but that structure is purely beneficial in the 99.99% case. You have workarounds which are arguably more idiomatic than what you are currently doing. See second half of my https://github.com/nunit/nunit/issues/1885#issuecomment-268680403 for examples.

**Comment by jnm2** [comment]
@CharliePoole 

> I'm a little lost in the part about "custom assembly builders."

Custom assembly builders addresses @oznetmaster's unique situation since we know he's not using the default one which puts parameters in `TestContext.Parameters`.

**Comment by oznetmaster** [comment]
Except that I do not fall into that category.

I am 100% compatible with the existing TestContext.Parameters property.
I am 100% compatiible with the existing CLI.

Everything else is added value.   However, it relies on passing the an IList in TestPackage.Settings["TestParameters"] to achieve that.  Passing an IDictionary<string, string>instead would be possible if I want to remove  some capabilities (null key as one simple example, but perhaps that could be coerced into String.Empty, which might work?), or to move the semantics of some capabilities back to the front end (CLI in my case) (the aggregation of an array from multiple identically names parameters, for example).

**Comment by oznetmaster** [comment]
Maybe the CLI, and VSTest, and the GUI, should all do all of this parsing, aggregation, and validation?  That would be one decision, I guess?

**Comment by jnm2** [comment]
@CharliePoole I also thought of adding a `"TestParametersDictionary"` key and disliked the duplication, but now I understand why it's important. 

Should we need to support both new framework with old front-end and old framework with new front-end?

|               | Old front-end                                               | New front-end                                                 |
|---------------|-------------------------------------------------------------|---------------------------------------------------------------|
| **Old framework** | (Front-end provides only string, old framework parses string) | Front-end provides string + dict, old framework parses string |
| **New framework** | Front-end provides only string, new framework parses string | Front-end provides string + dict, new framework copies dict   |

**Comment by jnm2** [comment]
> However, it relies on passing the an IList in TestPackage.Settings["TestParameters"] to achieve that. Passing an IDictionary<string, string>instead would be possible if I want to remove some capabilities (null key as one simple example, but perhaps that could be coerced into String.Empty, which might work?), or to move the semantics of some capabilities back to the front end (CLI in my case) (the aggregation of an array from multiple identically names parameters, for example).

Such is the nature of keeping hard forks up to date. If you change your command line parameter format and your test assembly builder adds a tiny amount of parsing to the the dictionary `{ [""] = "1|2|3" }`, you'll be able to continue business as usual.

This way we have good separation of concerns. The front-end is able to style the parameters to be parsed any way it likes or needs, and your test assembly builder only adds minimal parsing on top of that on your own level of abstraction. Seems perfect to me if I was in your shoes.

**Comment by oznetmaster** [comment]
I actually hate that syntax for arrays, and would never implement it. But that is personal, I guess.

And that does not address the issue of parameter aggregation.

Your approach will require changes to at least both CommandLineOptions (nunitlite and nunit.console), since both will now have to fully parse, and escape, the parameter option.  Seems needlessly redundant to me.

**Comment by oznetmaster** [comment]
Also, how does any front end provide an unnamed parameter in the IDictionary approach.  Apparently, that will never be an option in the future?

Also, are you planning on specifying the case sensitivity of the IDictionary, or is that up to the front end?

**Comment by oznetmaster** [comment]
I am fine with @CharliePoole approach, since it will stop this discussion. @jnm2 can use the new key for his needs, and the rest of us can continue to use the existing one for ours.

**Comment by oznetmaster** [comment]
This all started over the semi-colon, and escaping it.

In fact, after all of this discussion, that has not been solved.

The CLI (both of them) still have to handle the parsing and escaping, as they see fit.

**Comment by jnm2** [comment]
> I actually hate that syntax for arrays, and would never implement it. But that is personal, I guess.

That's neither here nor there. You can customize it. And if a front-ends parsing its own way to IDictionary is too restrictive, you'll have to fork the front-end which in the case of nunitlite I think you've already done.

> And that does not address the issue of parameter aggregation.

I'm not sure what you're referring to.

> Your approach will require changes to at least both CommandLineOptions (nunitlite and nunit.console), since both will now have to fully parse, and escape, the parameter option. Seems needlessly redundant to me.

I don't believe so. I believe that will still be `string` or `IList<string>` and won't be affected by this work. What we'll have is parsing inserted here between CommandLineOptions and the settings: https://github.com/nunit/nunit/blob/master/src/NUnitFramework/nunitlite/TextRunner.cs#L364

> Also, how does any front end provide an unnamed parameter in the IDictionary approach. Apparently, that will never be an option in the future?

Yes, as I've demonstrated several times now, `--params ""=1` or `--params =1` can easily be done, returning dictionary `{ [""] = "1" }`. It's up to your customization to decide what `""` means. Not hard.

> Also, are you planning on specifying the case sensitivity of the IDictionary, or is that up to the front end?

For everyone using DefaultTestAssemblyBuilder, they have no choice but to follow TestContext.Parameters which is case-sensitive. Therefore, all front-ends should create dictionaries this way.

**Comment by jnm2** [comment]
@CharliePoole I don't want the VSTest front-end to have to construct a string in the first place, so that results in this grid:

|               | Old VSTest front-end                                               | New VSTest front-end                                                 |
|---------------|-------------------------------------------------------------|---------------------------------------------------------------|
| **Old framework** | (Front-end provides only string, old framework parses string) | Front-end provides only dict, old framework sees no parameters |
| **New framework** | Front-end provides only string, new framework parses string | Front-end provides only dict, new framework copies dict   |

**Comment by oznetmaster** [comment]
@jhamm https://github.com/nunit/nunit/blob/master/src/NUnitFramework/nunitlite/TextRunner.cs#L364 is for nunitlite, but the same parsing has to occur in nunit.console as well.  Duplication of code.

**Comment by oznetmaster** [comment]
@jnm2 Windows command lines, are, for the most part, case insensitive.  File names are case insensitive, for instance. Most windows console programs accept options in both upper and lower case, as well as their values.

**Comment by oznetmaster** [comment]
`var result = String.Join (";"dictionary.Select(de => de.Key + "=" + de.Value.Replace (";", "\\;")).ToArray())`

That was really difficult :(

That was an approximate statement off the top of my head.

**Comment by oznetmaster** [comment]
This discussion has gotten way too ephemeral.

I think that @CharliePoole  had the right suggestion, and that it should apply universally.

**Comment by CharliePoole** [comment]
@jnm2 Commenting on your two tables, one for the command-line runners and one for the adapter, showing old/new front end against old/new framework. (Where old is taken to mean an older 3.x framework or runner.

1. I agree with the first table. Note that this means we have to solve the semicolon problem in order to do it.

2. I disagree with the cell in the second table that shows a breaking change. I don't think it's acceptable for a runner or adapter to require a minimal framework version other than 3.0. This is what we refer to as "backward" compatibility in the framework. All NUnit runners are supposed to be able to run tests built against any older framework version. That's because people have to recompile software in order to upgrade the framework but can use a new runner just by running it.

I would say that the adapter has to work just like the console or gui, providing both sets of values. I understand that it's an extra job that probably could have been avoided, but we can only work from the point where we now find ourselves. An alternative, of course, is to have this potential duplication sited in the engine, which is used by all runners except nunitlite.

-----

So now that we have talked all over the map, what are we going to actually implement? The present issue is not about the adapter, but the framework. We will have to make changes in this repo as well as in the console runner, adapter, gui and dotnet runner. What's the right order of doing things, bearing in mind that each component is released independently? Which are you volunteering for. 😄

**Comment by CharliePoole** [comment]
Just chiming in on case sensitiivity of options. This is a topic that always brings up strong feelings when it comes up, which is about every four or five years.

NUnit already decided long ago that command-line options are case sensitive. OTOH, we decided that literal values of optons could be case-insensitive. So we accept, unless something is broken, --process:Multiple, --process:multiple or even --process:MuLtIpLe but not --Process:Multiple

For the names of test parameters, which become the keys to a dictionary, case-sensitivity is currently required. I think it needs to stay that way.

**Comment by CharliePoole** [comment]
@jnm2 Just a side comment on the fact that we take the dictionary supplied by VS and create a semicolon-delimited string out of it... I can see how it would seem rather insane at first glance, but you might consider...

1. The test adapter isn't the main show for us, more of a side show.

2. It's probably the least absurd of the absurdities we perpetrate when trying to shoehorn VS concepts about what a test is into the NUnit framework point of view. Once a decision is made on how we want to escape semicolons, the current code that builds a string can be trivially changed to do it.

----

Given that a string version has to be provided anyway, are you still convinced that a separate dictionary is worth the effort? What do we get for it?

**Comment by oznetmaster** [comment]
My example showed escaping semi-colons using industry standard escape notation.  No reason to use anything else that I can see.

**Comment by oznetmaster** [comment]
I also agree with @CharliePoole .  I do not see the advantage in the IDictionary.

**Comment by oznetmaster** [comment]
@CharliePoole I did not realize that was a conscious choice about the case of command line options.

I have always made the ones that I added case-insensitive.  I think the CF build even allows others to be as well.  Guess I was bad :(

**Comment by CharliePoole** [comment]
@oznetmaster I assume that by industry standard you mean "\;"? I'm fine with that, I just wanted somebody to say what it would be - preferably whoever is going to write the code.

Regarding IDictionary (or IList), I'll say that @jnm2 has me convinced it __would have__ been a better approach in the first place, since it allows a runner that already has structured data to supply it. At the point where we are and given that we have to remain backward compatible, I'm not sure there is a lot of incremental benefit to allow clients to supply the data in a structured way, but maybe @jnm2 sees something.

Yes... not only was it a conscious choice - inspired in part by how badly case-insensitivity sometimes plays when ported to linux - but it has been revisited in a major way at least three times during my time with this project! However, I do not believe you have deviated from it in your nunit contributions. That would have required modifying the underlying Options.cs file we got from mono, which assumes a case-insensitive approach.

**Comment by oznetmaster** [comment]
No, I meant exactly what I said.  To escape a semicolon, use `"\;"` and to escape a backslash use '"\\"`

No, mono assumes a case sensitive approach for option names, as you indicated.  In my build, I put both cases in the list of allowable option names, so that, for example, both --params and --Params are allowed.

**Comment by jnm2** [comment]
@CharliePoole 

> I agree with the first table. Note that this means we have to solve the semicolon problem in order to do it.

So the VSTest adapter will continue building the string exactly as it does today with no escaping, because if it started escaping it would be a backward-breaking change, but it will also provide a dictionary in case the framework is new enough.

> So now that we have talked all over the map, what are we going to actually implement? The present issue is not about the adapter, but the framework. We will have to make changes in this repo as well as in the console runner, adapter, gui and dotnet runner. What's the right order of doing things, bearing in mind that each component is released independently? Which are you volunteering for.

I'm volunteering for all of it. The order genuinely doesn't matter due to our awareness of back and forward compat. In the order that would benefit me:

 1. Framework: if `"TestParametersDictionary"` is present, it will be copied to `TestContext.Parameters` and the `"TestParameters"` string will be ignored. This can ship any time.
 2. VSTest adapter: add a `"TestParametersDictionary"` from the .xmlsettings before continuing to adding the `"TestParameters"` string for compat. This can ship any time.

Things I don't care about (yet) but am willing to contribute:

 3. NUnitLite, console runner, and dotnet runner: parse CLI params to dictionary and put it in `"TestParametersDictionary"` for the framework. Each of these can ship any time.
    We should discuss parsing code duplication. The only downside is that we're initially copying the parsing code from the old framework and copying it to these three places. One of the wins here is that each CLI can do what makes the most sense, e.g. if the dotnet runner wants to follow dotnet CLI conventions. In the future, different CLIs might want to provide a way to escape semicolons, or they might stop separating via semicolons and require `--p` to be specified multiple times.

**Comment by jnm2** [comment]
@CharliePoole

> Regarding IDictionary (or IList), I'll say that @jnm2 has me convinced it would have been a better approach in the first place, since it allows a runner that already has structured data to supply it. At the point where we are and given that we have to remain backward compatible, I'm not sure there is a lot of incremental benefit to allow clients to supply the data in a structured way, but maybe @jnm2 sees something.

I do. We can't touch `"TestParameters"`, not even to escape semicolons, without breaking the old-new and new-old corners of the compatibility table. We absolutely have to add a second key, or else we accept breaking.

Once we have decided not to break and therefore to provide a new key, it is a new start and we may as well do what we'd prefer to do with a fresh start and use `IDictionary`.

**Comment by jnm2** [comment]
Say we change the new framework and the new runner to escape with `\;` which seems the most obvious possible escaping mechanism, which also requires escaping `\` as `\\`:
 * New runner with old framework will turn user-input `X=1;` into TestContext `X=1\` and user-input `C:\my\path\` into TestContext `C:\\my\\path\\`
 * Old runner with new framework will turn user-input `\\ test` into TestContext `\ test`

**Comment by CharliePoole** [comment]
OK @jnm2 I see your point now. I was thinking that fixing a bug is not a breaking change. However, the obvious escaping mechanism using backslash will break existing working cases that contain a backslash in the value. It seems as if there is no way to completely fix this in only one place.

I believe a partial fix is possible, in the framework itself, provided __no__ escaping has been done. It would have to be based on the idea that a "normal" parameter string that is passed to the framework will have some number 'n' of `=` characters and `n-1` semicolons. Any extra (leading) semicolons that occur between two equal signs should be treated as data. Any extra (following) equal signs should be treated as data. Values that contain both `;` and `=' would be broken, but they are broken now. Given a good set of test data, I think I could make this work for most cases.

That leaves us with three ways to proceed, I think.

1. Adding the completely resolved dictionary or list makes the most sense for the long term with both new framework and new driver. It puts the burden of parsing on each client, but we accepted that sort of functional duplication when we decided to break nunit into separately released components.

2. Adding escaping to each client and unescaping to the framework could also fix the new/new combination.

3. Adding a more sensitive interpretation of the parameter string to the framework could fix some of the problems in the old runner / new framework combination - perhaps the most common ones.

We can combine these as 1+2 or 1+3. The 2 + 3 combination would require a closer look.

Here's a question for the @nunit/core-team - it's exactly the kind of thing where I hoped that group would intervene in a project and particularly in a cross-project issue...

> We have figured out that the implementation of run parameters is broken. The feature was only just released in 3.5 and doesn't work for all cases. If this had been out for several releases, we would strongly want a backward compatible solution. How strongly should we push for backward compatibility for a feature that has only been out for one release?

Meanwhile I hope that @jnm2 will proceed with approach number 1, which can be handled separately and that both @jnm2 and @oznetmaster will send me or post test cases for approach number 3.

@jnm2 I'm assigning this issue to you. Is there already an nunit 3 adapter issue? If not there should be.

**Comment by oznetmaster** [comment]
Actually, @CharliePoole is not totally correct in saying "the obvious escaping mechanism using backslash will break existing working cases that contain a backslash in the value".

It will only break existing working cases if the backslash occurs immediately before the terminating semicolon for that value, a significantly less likely possibility.

I currently use the following code to parse an parameter definition string, which will break as above (but no one has ever complained):

```
			this.Add ("params|p|Params|P=", "Define a test parameter.",
				 v =>
					 {
					 string parameters = RequiredValue (v, "--params");
					 parameters = parameters.Replace ("\\\\", "\x01").Replace ("\\;", "\x02").Replace (';', '\x00').Replace ('\x01', '\\').Replace ('\x02', ';');
					 foreach (string param in parameters.Split (new[] { '\x00' }))
						 {
						 if (!param.Contains ("="))
							 ErrorMessages.Add ("Invalid format for test parameter. Use NAME=VALUE.");
						 if (testParameters == null)
							 testParameters = new List<string> ();
						 testParameters.Add (param);
						 }
```

where

```
		private List<string> testParameters;
```

**Comment by jnm2** [comment]
@oznetmaster You forgot about the necessity of escaping `\`. If you had read my examples, you'd see why this would cause issues:

 * New runner with old framework will turn user-input `\my\path` into TestContext `\\my\\path`
 * Old runner with new framework will turn user-input `\\server\share\folder` into TestContext `\server\share\folder`

**Comment by oznetmaster** [comment]
@jnm2 You obviously did not read the code very closely.  Escaped backslashes are properly handled by that code.

My entire discussion above was about how backslashes cause breaking issues when escaping was added.

**Comment by oznetmaster** [comment]
I also indicated that this is the code my CF build uses.  The escaping is done in the CLI, and not in the framework in that build.

**Comment by jnm2** [comment]
@oznetmaster I actually ran your code in LinqPad and it turned `@"path=\\server\share\path"` into `new List<string> { @"\server\share\path" }`. I must be misunderstanding something.

Remember we have to deal with all four corners of the compatibility table. Your `Escape` method (which is technically unescaping here) only works if the runner escapes `\` to `\\`, but NUnit's new framework release has to support old runners that do not escape `\` to `\\` and expect `\\` to stay as `\\`.

**Comment by oznetmaster** [comment]
Yes, that is one of the breaking escapes of the backslash.  It would have to be written (on the command line) as path=\\\\server\\share\\path to be correctly escaped (or path=\\\\server\share\path).  However, single backslashes that are not before a semicolon will still work properly (as above), since only backslash and semicolon are escapable.

Again, as far as I am concerned, since this is a command line only issue, a "breaking" change is minor, especially in the CF build.

As I noted, not a single user has even mentioned the breaking change to me.  I assume they just use the correct escaping when typing there command lines.

Actually, going back over my "bug" reports, I had two previous reports that complained that normal c# escaping did not work (before the "fix").

However, one question that this raises is if full escaping should be supported, so that control characters and special characters could be easily included in the parameter value.  I have not done this, but it, in retrospect, seems reasonable.

**Comment by oznetmaster** [comment]
Also, if you reread the post, you will note I only posted that code to show how I chose to "break" compatibility with the previous CLI when I fixed the semicolon bug.  I was certainly not proposing this as a solution to the general problem, for which I am not convinced there is a non-breaking solution, only come-closer ones.

**Comment by jnm2** [comment]
> As I noted, not a single user has even mentioned the breaking change to me. I assume they just use the correct escaping when typing there command lines.

Remember that `\\server\share\path` may not be coming from a command line but from VSTest or being put in the command line by Cake or some other build script.

**Comment by CharliePoole** [comment]
Even so, I like @oznetmaster 's approach to escaping, which I hadn't looked at before. Perhaps it can be enhanced, but even if not it's the closest approximation I have seen.

@oznetmaster just as a comment, I'm afraid you overestimate our ability to keep up with all these comments. :smile: I hadn't noticed that code before. Also, I edited your comment to put the examples with mutliple backslashes between back quotes in order to clarify.

It seems to me that one could improve the outcome slightly by __only__ dealing with backslash before semicolon, as follows:

```C#
parameters = parameters.Replace ("\\\\;", "\x01").Replace ("\\;", "\x02").Replace (';', '\x00').Replace ('\x01', '\\;').Replace ('\x02', ';');
```

Users would be told to type `\;` for an embedded semicolon and `\\;` in the extremely unlikely event that they wanted an actual backslash before a semicolon.

I think separating this issue into two parts, one dealing with escaping and unescaping and the other using a dictionary provided by the runner makes things a lot easier.

**Comment by oznetmaster** [comment]
Actually, the backslash before the semicolon is more common the you might think.  `path=c:\folder1\folder2\`, for instance.

However, I agree with @CharliePoole that his approach does increase "backward" compatibility.

If my issue #1936 is considered, however, then the need to escape the escape character is global.

**Comment by CharliePoole** [comment]
Here's something that @oznetmaster suggested once before... should we consider disallowing the use of multiple, semicolon-separated parameters on a single option? Rather than typing `-p:p1=X;p2=Y` the user would have to do `-p:p1=X -p:p2=Y`. It seems less convenient for the user but it does entirely eliminate any ambiguity in the use of semicolons and backslashes. It's a breaking change in a different sense: scripts will have to be modified. But this is less of an issue than other kinds of breaks, IMO.

**Comment by jnm2** [comment]
@CharliePoole For the existing `"TestParameters"` string would break our compat table again, and going forward we'll have IDictionary so runners can make their own decisions about this.

**Comment by jnm2** [comment]
Old VSTest runner forces the format `p1=X;p2=Y`, and the new framework interprets it as `{ ["p1"] = "X;p2=Y" }`.

**Comment by oznetmaster** [comment]
I still think disallowing the use of multiple, semicolon-separated parameters in a single option certainly solves a problem.

Even if we allow full escaping for special characters, this still makes it unambiguous.

**Comment by jnm2** [comment]
If we are going to break compat, let's be blunt about it (no params because IDictionary not string) and not subtle (`{ ["p1"] = "X;p2=Y" }` or things with `x=path\;y=other\path\` turning to `{ ["x"] = @"path\;y=other\path\" }`).

**Comment by CharliePoole** [comment]
@jnm2 My comment had to do with typing options at the command line - no relation to the VS adapter and no impact at all on the compatibility table. 

In your last suggestion, are you saying that we should __not__ provide the string parameter at all from the new runners?

**Comment by jnm2** [comment]
@CharliePoole 

> My comment had to do with typing options at the command line - no relation to the VS adapter and no impact at all on the compatibility table.

As long as we don't change the semicolon behavior in the framework. But then what did you mean by "disallow"? What happens if the user types `-p:p1=X;p2=Y` anyway? I don't see any way to disallow that when running against the old framework without breaking.

> In your last suggestion, are you saying that we should **not** provide the string parameter at all from the new runners?

**Only** if we are causing any kind of breaking change, then yes. **If** a breaking change is make, which so far any alteration to the string parsing seems to be, better to make it an obvious break like parameters not working at all or a message about version incompatibility than a subtle break with perhaps worse consequences. 
Similar to how you want to throw if an argument is null even if it won't end up being used in this particular call, so that people realize they're using it wrong the first time they call and there are no surprises.

**Comment by jnm2** [comment]
@CharliePoole is there anything else that you would like me to do here?

**Comment by CharliePoole** [comment]
I'm closing this although I think there may be other things to do afterwards. However, this particular issue has gotten incredibly cluttered, so lets do new, separate issues for anything else.

We now are using a dictionary for the run settings and have put the responsibility in the right places. But it seems to me that what may be missing is the __actual__ dealing with semicolons embedded in the attributes, which is what started the entire discussion.

**Comment by CharliePoole** [comment]
@jnm2 You have reopened a number of issues? Is it a mistake? If not you should comment each one.

**Comment by jnm2** [comment]
I must have done something wrong in ZenHub. I'm sorry!

**Comment by andrewliang** [comment]
Is there any chance this fix can be applied NunitLite's console? 
Thanks!

**Comment by CharliePoole** [comment]
Based on memory and reading the comments above, this fix __was__ for NUnitLite.

The issue was closed two years ago and the fix released with NUnit 3.7. Of course, we could have a regression now, in which case a new issue with a description of the problem would get us going on it.

**Comment by jnm2** [comment]
@andrewliang Are you asking if the new `--testparam` changes can be ported from https://github.com/nunit/nunit-console/pull/575 to NUnitLite?
